Add GooglePreprocessor and SmartPreprocessor

vemel opened the pull request:
GooglePreprocessor - supports Google format for docstrings
SmartPreprocessor - wisely chooses between RST and Google preprocessors for each docstring

NiklasRosenstein commented on pydocmd/loader.py:
```diff
@@ -31,117 +31,122 @@
+            new_line = ' ' * (delta - indent) + new_line
+        lines[i] = new_line
+
+    return '\n'.join(lines)
```
Did the line terminator change here?

NiklasRosenstein replied:
Ah no the indentation changed. Please go back to two-space, I know it's not common but the rest of the source is currently using a two-space indentation. I'd rather have this PR contain only the changes it intends to address, and not _also_ try to perform cleanup

vemel replied:
Will do

NiklasRosenstein commented on pydocmd/preprocessors/smart.py:
```diff
@@ -0,0 +1,38 @@
+
+class Preprocessor(object):
+    """
+    This class implements the preprocessor for restructured text and google.
```
I like the idea of a smart preprocessor!

vemel replied:
Works for my project. I did not add simple preprocessor as there is a call of `link_lookup` method that is missing. I could not find it, so I could not test simple preprocessor

vemel replied:
Reverted pylint auto formatting, now it should be easier to see what has changed

NiklasRosenstein replied:
Thanks! The link_lookup got introduced in another PR and is a bit ugly at the moment. I believe you would just need to propagate the attribute from the smart processor object to the actual processor that will be used

NiklasRosenstein commented:
Please also be aware that I'm working on a new version that parses the code with `lib2to3` instead of executing the code on the `develop` barnch. It's actually not far from being merged into master.

Anyway, happy to merge your PR into the 2.x code base. I'll likely copy the processor code to the 3.x branch afterwards.

vemel commented:
Rebased to master, should be no conflicts now.

NiklasRosenstein commented on pydocmd/__main__.py:
```diff
@@ -39,244 +39,254 @@
 
```
Could you please revert these indentation changes as well?

vemel replied:
Done

NiklasRosenstein commented:
Please take a look at the [`pr-84`](https://github.com/NiklasRosenstein/pydoc-markdown/tree/pr-84) branch. I've added some unit tests. Could you take a look at why the smart preprocessor does not detect the Google syntax properly? 😃

vemel commented:
I will take a look, thanks! Actually, I implemented GooglePrerocessor incorrectly, it also supports https://www.python.org/dev/peps/pep-0257 we should probably split these two guys, and we can create a parent class for them as methods will be the same, just param regexps should be different.

FOr now logic for detecting docstring style is quite simple. If I find a section from Google/PEP-0257 - 
 (e.g. `Args:`) I use GooglePreprocessor. Otherwise, RST is used.

vemel commented:
> Please take a look at the [`pr-84`](https://github.com/NiklasRosenstein/pydoc-markdown/tree/pr-84) branch. I've added some unit tests. Could you take a look at why the smart preprocessor does not detect the Google syntax properly?

Fixed, regexp was incorrect.
